src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,19 +36,216 @@ case class NoOpFingerprint() {
+   * @return Content object with Directory tree from leaf files.
+   */
+  def fromLeafFiles(files: Seq[FileStatus]): Directory = {
+    require(
+      files.nonEmpty,
+      s"Empty files list found while creating a ${Directory.getClass.getName}.")
+
+    require(
+      files.forall(!_.isDirectory),
+      "All files must be leaf files for creation of Directory.")
+
+    /* from org.apache.spark.sql.execution.datasources.InMemoryFileIndex. */
+    val leafDirToChildrenFiles = files.toArray.groupBy(_.getPath.getParent)
+
+    // Hashmap from directory path to Directory object, used below for quick access from path.
+    val pathToDirectory = HashMap[Path, Directory]()
+
+    for (path <- leafDirToChildrenFiles.keys) {
+      val files = ListBuffer[FileInfo]()
+      files.appendAll(leafDirToChildrenFiles.getOrElse(path, Array()).map(FileInfo(_)))
+
+      if (pathToDirectory.contains(path)) {
+        // Map already contains this directory. Just append the files to its existing list.
+        pathToDirectory(path).files.asInstanceOf[ListBuffer[FileInfo]].appendAll(files)
+      } else {
+        var location = path
+        // Create a new Directory object and add it to Map
+        val subDirs = ListBuffer[Directory]()
+        var directory = Directory(location.getName, files = files, subDirs = subDirs)
+        pathToDirectory.put(location, directory)
+
+        // Keep creating parent Directory objects and add to the map if non-existing.
+        while (location.getParent != null && !pathToDirectory.contains(location.getParent)) {
+          location = location.getParent
+          if (location.isRoot) {
+            directory = Directory(
+              location.toString,
+              subDirs = ListBuffer(directory),
+              files = ListBuffer[FileInfo]())
+          } else {
+            directory = Directory(
+              location.getName,
+              subDirs = ListBuffer(directory),
+              files = ListBuffer[FileInfo]())
+          }
+          pathToDirectory.put(location, directory)
+        }
+
+        // Either root is reached (parent == null) or an existing directory is found. If it's the
+        // latter, add the newly created directory tree to its subDirs.
+        if (location.getParent != null) {
+          pathToDirectory(location.getParent).subDirs
+            .asInstanceOf[ListBuffer[Directory]]
+            .append(directory)
+        }
+      }
+    }
+
+    val rootPath = getRoot(leafDirToChildrenFiles.head._1)
+
+    pathToDirectory(rootPath)
+  }
+
+  // Return file system root path from any path. E.g. "file:/C:/a/b/c" will have root "file:/C:/".
+  // For linux systems, this root will be "file:/". Other hdfs compatible file systems will have
+  // corresponding roots.
+  @tailrec
+  private def getRoot(path: Path): Path = {
+    if (path.isRoot) path else getRoot(path.getParent)
+  }
+
+  // Create Directory from `dirPath` where `dirPath` is the directory containing
```

Comment:
This is a bit confusing since this doesn't populate the "root" path, and it's being handled by the caller.

I haven't tested but, can't you do something like the following?
```scala
  private def createDirectory(dirPath: Path, leafFiles: Array[FileStatus]): Directory = {
    // Recursively creates parent directory to have the given `subDir`.
    @tailrec
    def createParentDirectory(dirPath: Path, subDir: Directory): Directory = {
      if (dirPath.isRoot) {
        Directory(dirPath.toString, files = Seq(), subDirs = Seq(subDir))
      } else {
        createParentDirectory(
          dirPath.getParent,
          Directory(dirPath.getName, files = Seq(), subDirs = Seq(subDir)))
      }
    }

    val files = leafFiles.map(FileInfo(_))
    if (dirPath.isRoot) {
      Directory(dirPath.toString, files = files)
    } else {
      createParentDirectory(dirPath.getParent, Directory(dirPath.getName, files = files))
    }
  }
```

And the caller simply becomes:
```scala
if (leafFiles.nonEmpty) {
  fromLeafFiles(leafFiles)
} else {
  createDirectory(path, Array())
}
```

I could be missing something, but if this doesn't work, let's make sure this function doesn't handle the "Root"

Reply:
thanks @imback82 , I cleaned up this section of the code so most of it is not required anymore (it came to me from your other comment where you mentioned `createParenteDirectory` is being called only with empty `Array()`.